### ISSUE TITLE
Attempting to limit response size when calling upstream server

### ISSUE THREAD

**Issue by swav** [envoyproxy/envoy#17340]
I have configured envoy (v1.18.3) as a dynamic forward proxy to serve as a proxy to load image files from external, third-party servers. This works fine. The additional requirement I have, is to limit the size of images loaded from these servers to 5MiB, which I've attempted to do by specifying **per_connection_buffer_limit_bytes** set to 5242880 (5 MiB), but I see no effect of this setting at all. 
I've also attempted to specify http2 protocol options (not shown below) with initial_stream_window_size & initial_connection_window_size values but that was wrong as the protocol used by the third-party servers I was accessing was htttp 1.1 so I ended up getting 502 response codes.

So, with the sample configuration shown below, is there something else I need to do achieve my goal. I've assumed that the limits need to be done in my cluster definition as this is the upstream side of the call. I followed this entry in the docs regarding the flow control: https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/flow_control

thanks in advance,
swav

"dynamic_active_clusters": [
    {
     "version_info": "1626268857502264421",
     "cluster": {
      "@type": "type.googleapis.com/envoy.config.cluster.v3.Cluster",
      "name": "dynamic_forward_proxy_cluster",
      "connect_timeout": "1s",
      "per_connection_buffer_limit_bytes": 5242880,
      "lb_policy": "CLUSTER_PROVIDED",
      "transport_socket": {
       "name": "envoy.transport_socket.tls",
       "typed_config": {
        "@type": "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext",
        "common_tls_context": {
         "validation_context": {
          "trusted_ca": {
           "filename": "/etc/ssl/certs/ca-certificates.crt"
          }
         }
        }
       }
      },
      "cluster_type": {
       "name": "envoy.clusters.dynamic_forward_proxy",
       "typed_config": {
        "@type": "type.googleapis.com/envoy.extensions.clusters.dynamic_forward_proxy.v3.ClusterConfig",
        "dns_cache_config": {
         "name": "dynamic_forward_proxy_cache_config",
         "dns_lookup_family": "V4_ONLY"
        }
       }
      }
     },

**Comment by lambdai** [comment]
Per_connection_buffer_limit_bytes is applied on buffer of tcp connection. It might applies to http1 stream but not on http2 stream. IIRC The latter by default is 256 MB.
You are on the right track to set http2 options, but you may also want to set auto http protocol on upstream.

It seems you should use the [http_protocol_options upstream extension](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/upstreams/http/v3/http_protocol_options.proto#extensions-upstreams-http-v3-httpprotocoloptions
) in your [cluster config](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto.html?highlight=envoy.upstreams.http.http_protocol_options)

**Comment by snowp** [comment]
I don't think we have a way to limit how big the response is out of the box, though it might be fairly easy to do with a lua or C++ filter that checks the value of content-length and stands in a failure response if its over the threshold.

**Comment by swav** [comment]
@snowp Thanks for the suggestion. I was thinking about content-length header in lua script, so this confirms my approach.

**Comment by swav** [comment]
I have a follow up to the question raised originally. So, I have implemented lua on_response filter to read value of Content-Length response header and if it exceeds some predefined value, in my case 5MiB, to write new response body with an error message, like this:
if errMsg ~= nil then
  local content_length = response_handle:body():setBytes(errMsg)
  response_handle:headers():replace("content-length", content_length)
end
This works as expected but if I attempt to request an image file which will exceed 10MiB in size, I receive a 500 status code in response. 
The log shows this: Sending local reply with details response_payload_too_large directly to the encoder

So, I found that the way to allow response bodies larger than 10MiB is to add 
'per_connection_buffer_limit_bytes: 12582912' (12MiB) on cluster AND listener level. I've noticed that just setting it on cluster level had no effect.
This brings me to these questions:
1) what is the relationship between per_connection_buffer_limit_bytes set on cluster level vs on listener level?
2) why setting it only on cluster level only has no effect - I would prefer to be able to only affect a single cluster with this setting, as opposed to entire listener, which has many more routes and I don't want to limit any other response sizes only a single one
3) in the above scenario, when my lua script determines that response is larger than my own pre-configured limit (1 MiB) but response is larger than what I set as per_connection_buffer_limit_bytes value (3 MiB) why do I get only 500 error and my own 502 response status and my own custom response header get lost?

I've attached my envoy config yaml (as txt) file, the lua filter dealing with response size is: responselimit.lua .
[envoy-config-host-rewrite-1.18.3.txt](https://github.com/envoyproxy/envoy/files/6934547/envoy-config-host-rewrite-1.18.3.txt)

thank you, in advance
